src/Storages/StorageURL.cpp

```diff
@@ -271,6 +206,72 @@ namespace
             }
         }
 
+        static std::unique_ptr<ReadBuffer> getFirstAvailableURLReadBuffer(
+            const std::vector<String> & urls,
+            ContextPtr context,
+            const URIParams & params,
+            const String & http_method,
+            std::function<void(std::ostream &)> callback,
+            const ConnectionTimeouts & timeouts,
+            const String & compression_method,
+            const ReadWriteBufferFromHTTP::HTTPHeaderEntries & headers,
+            bool skip_url_not_found_error)
+        {
+            Poco::Net::HTTPBasicCredentials credentials{};
+            String first_exception_message;
+
+            for (auto option = urls.begin(); option < urls.end(); ++option)
+            {
+                auto request_uri = Poco::URI(*option);
+                for (const auto & [param, value] : params)
+                    request_uri.addQueryParameter(param, value);
+
+                try
+                {
+                    std::string user_info = request_uri.getUserInfo();
+                    if (!user_info.empty())
+                    {
+                        std::size_t n = user_info.find(':');
+                        if (n != std::string::npos)
+                        {
+                            credentials.setUsername(user_info.substr(0, n));
+                            credentials.setPassword(user_info.substr(n + 1));
+                        }
+                    }
+
+                    return wrapReadBufferWithCompressionMethod(
+                        std::make_unique<ReadWriteBufferFromHTTP>(
+                            request_uri,
+                            http_method,
+                            callback,
+                            timeouts,
+                            credentials,
+                            context->getSettingsRef().max_http_get_redirects,
+                            DBMS_DEFAULT_BUFFER_SIZE,
+                            context->getReadSettings(),
+                            headers,
+                            ReadWriteBufferFromHTTP::Range{},
+                            context->getRemoteHostFilter(),
+                            /* delay_initiliazation */urls.size() == 1,
+                            /* use_external_buffer */false,
+                            skip_url_not_found_error),
```

Comment:
But if we set `skip_url_not_found_error=true` for all options we will break the next scenario:
Imagine we have 2 url options: `option1, option2` and url `option1` doesn't exists but url `option2` exists and contains the data. In this case we won't throw an exception from `ReadWriteBufferFromHTTP` constructor with `option1` since `skip_url_not_found_error=true` and we return a read buffer from url `option1` and we won't read any data from it though we have data in url `option2`. I think we should set `skip_url_not_found_error` only for the last option here.

Reply:
Now in case of many options we set `skip_url_not_found_error=false` for all options, but we want to be able to skip them all if they are all don't exist, so let's set `skip_url_not_found_error=true` for the last option.